### ISSUE TITLE
Add "NEW" badge for GA4 metrics

### ISSUE THREAD

**Issue by nfmohit** [google/site-kit-wp#6552]
## Feature Description

For the GA4 metrics rendered in the Site Kit dashboard, we will need to add a "NEW" badge to each of them according to the [Figma designs](https://www.figma.com/file/0UCIE3hPeiu4NYdpAODASd/GA4-Updates?node-id=1%3A1168&t=ckvN59ELFbj9QQWN-0). A tooltip with explanatory text should appear upon hovering on the "NEW" badges, according to the [Figma designs](https://www.figma.com/file/0UCIE3hPeiu4NYdpAODASd/GA4-Updates?node-id=171%3A3012&t=ckvN59ELFbj9QQWN-0).

The challenge here is to have a clickable link inside the tooltip, which will disappear as soon as the cursor is moved from the badge. There have been discussions around this in Figma [here](https://www.figma.com/file/0UCIE3hPeiu4NYdpAODASd?node-id=171:3012#346337032).

[Here](https://docs.google.com/document/d/14pCNmdPBqAJ-RfPhbHXfaMhGaPK-n-7F-6Sgujd0XCQ/edit#heading=h.4ibds0gvmtor) is the relevant part in the design doc.

Here is a screenshot for visual aid:
![image](https://user-images.githubusercontent.com/20284937/217500308-2abc84d9-3edf-4d8c-8e49-2c36d08193be.png)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A "New" badge should appear next to any metric name that is sourced from data unique to Google Analytics 4. The Figma design shows where all "new" badges should be, but for reference, they are:
  *  Search Funnel "Conversions" metric/widget: 
![CleanShot 2023-03-07 at 19 04 12](https://user-images.githubusercontent.com/90871/223525033-18d4646f-7e24-48d6-bf8a-5a09a53fa504.png)

  * Popular Pages "Sessions" and "Engagement Rate" metrics (note, the design refers to "Engaged Sessions" but this has been updated to be "Engagement Rate":
![CleanShot 2023-03-07 at 19 04 48](https://user-images.githubusercontent.com/90871/223525117-a2a2a1c1-490a-4d73-a45d-3f4846192a05.png)

* When hovering over (tapping on mobile) a "New" badge, a tooltip should appear like in the Figma design:
<img width="512" alt="CleanShot 2023-03-07 at 19 19 45" src="https://user-images.githubusercontent.com/90871/223528381-06b7cae7-2256-4eec-acb4-4749fca6c4b7.png">

* The text in the tooltip should be:
> $METRIC_NAME is a new Google Analytics 4 metric replacing the Goals metric. <a>Learn more</a>
  * Where `$METRIC_NAME` is the name of the metric.
  * The "Learn more" link should be:
    * Conversion metric: https://support.google.com/analytics/answer/9267568
    * Sessions metric: https://support.google.com/analytics/answer/9191807
    * Engagement Rate metric: https://support.google.com/analytics/answer/12195621

## Implementation Brief

*  Within `assets/js/components` create a new `NewMetricBadge.js` component.
    * Use the `Badge` component to create a dedicated `NewMetricBadge` with a "New" title and tooltip content as per the AC.
    * Introduce new `tooltipTitle` and `learnMoreLink` props to configure the tooltip content.
    * Use the `Tooltip` component to render a configurable tooltip around`Badge` with the content.
    * Use the `Link` component to render a link in the tooltip as per the AC.
    * Use the `interactive` prop for the tooltip as per the [documentation](https://v4.mui.com/components/tooltips/#interactive) for the ability to click the link.
    * Wrap `Badge` in `forwardRef()` and spread its props into its `span` component, in order for `Badge` to be used as a `Tooltip` target, as per the [documentation](https://v4.mui.com/components/tooltips/#custom-child-element).
    * Introduce a new class and apply styles to `Badge` as per the [Figma designs](https://www.figma.com/file/0UCIE3hPeiu4NYdpAODASd/GA4-Updates?node-id=171-3164&t=umqvQCOYgQI7xYXD-4).
      * Add the styling to the new file `assets/sass/components/global/_googlesitekit-new-metric-badge.scss`.
   
* For a bit more context, and a concrete example of how the `NewMetricBadge` component could look, see this https://github.com/google/site-kit-wp/issues/6552#issuecomment-1474183079 below.

* For the `Conversion` metric, add configuration to the relevant entry in `dataBlocks` in `assets/js/modules/search-console/components/dashboard/SearchFunnelWidgetGA4/Overview/index.js`, adding a `badge` property which is a `NewMetricBadge` element.
    * In  `DataBlock` render the new `badge` property after `titles`.

* For the `Sessions` and `Engagement Rate` metrics, add configuration to the relevant entries in `tableColumns` in `assets/js/modules/analytics/components/module/ModulePopularPagesWidgetGA4.js`, adding a `badge` property which is a `NewMetricBadge` element.
    * In  `ReportTable` render the new `badge` property after `titles`.

* Add a story for the `NewMetricsBadge` component, and also add variants for the `Search Funnel` and `Popular Pages` widget stories that show the "New" badges.
* Add a story to `ReportTable` to include an example of the badge prop being used.

### Test Coverage

* Add VRT scenarios for the new stories.
* All tests should pass

## QA Brief

* The GA4 specific Conversions, Session and Engagement Rate metrics should have the `New` badge added to their titles in the GA4 version of Site Kit Dashboard. 
* Hovering the badges should open a tooltip with the copies present in the Figma design (Note that while the AC has a copy, it only makes sense for the Conversions metric, and the others were pulled directly from Figma.)
    * The learn more link should be clickable and open the links defined in AC in a new browser tab. 
* Ensure the design matches with Figma (compare the tokens). 

## Changelog entry

* Add badges to highlight new metrics on the Analytics 4 version of the dashboard.

**Comment by techanvil** [comment]
Hi @sashadoes, thanks for the IB here. I have a few comments on it:

- We already have a `Badge` component - it might be worth exploring this to see whether it can be extended and reused here, or alternatively, create a new component with a different name.
- It seems a bit strange to say we should "wrap their titles with `NewMetricBadge`", this makes it sound like the title should be passed into `NewMetricBadge` as a prop. I don't think this is your intent, did you mean to say something like "render the `NewMetricBadge` to the right of their titles"?
- It would be good to list all of the components to be updated rather than leave it to the implementer to determine these. This would be useful for implementation, and also better inform the estimate. Furthermore I'd suggest digging into these a bit to see how the new badge can be rendered, as there may be some additional changes required that aren't specced out here. For example it looks like the [`DataBlock`](https://github.com/google/site-kit-wp/blob/f04ed052ce6b654378eb1e41cfdbe76f774b7692/assets/js/components/DataBlock/index.js#L102-L109) will need updating as it currently expects a `string` for its `title` prop. We don't have to go into every last detail, but it would be good to get a more comprehensive picture of things.
- Although not explicitly stated in the AC, we need to accommodate this point that is mentioned in the description, in order to be able to click on the "Learn more" link. You can see this in action, for example by hovering over the user menu icon and then moving the cursor over the tooltip which will disappear.
> The challenge here is to have a clickable link inside the tooltip, which will disappear as soon as the cursor is moved from the badge. There have been discussions around this in Figma here.

I've also noticed the AC is a bit incomplete as the "Learn more" link provided is applicable to the Conversions metric, and we need to determine the links to use for the other metrics. @nfmohit, is this something you can help out with?

**Comment by nfmohit** [comment]
> I've also noticed the AC is a bit incomplete as the "Learn more" link provided is applicable to the Conversions metric, and we need to determine the links to use for the other metrics. @nfmohit, is this something you can help out with?

Thank you for the flag, @techanvil! I've updated the ACs to include links for all metrics. Thanks!

**Comment by techanvil** [comment]
> Thank you for the flag, @techanvil! I've updated the ACs to include links for all metrics. Thanks!

Top stuff, thanks @nfmohit!

**Comment by techanvil** [comment]
Hey @sashadoes, thanks for addressing those points. However, there's an aspect that still needs to be covered. I've dug into this myself and in doing so, have had a bit of a realisation about the implementation that makes me think we should take a slightly different approach.

So, as mentioned previously we still need to cover this point:

> Although not explicitly stated in the AC, we need to accommodate this point that is mentioned in the description, in order to be able to click on the "Learn more" link. You can see this in action, for example by hovering over the user menu icon and then moving the cursor over the tooltip which will disappear.

To further illustrate you can navigate to the [Icon Button with Tooltip](https://google.github.io/site-kit-wp/storybook/develop/?path=/story/components-button--icon-button) story and again you'll see the tooltip disappears when moving the cursor away from the button which means the tooltip content is not clickable.

It turns out the solution to this is simple. As per the [documentation](https://v4.mui.com/components/tooltips/#interactive) for the third-party `Tooltip` component we're using, we just need to add the `interactive` prop to the tooltip.

As mentioned at the start of this comment, looking into the `Tooltip` has made me reconsider the implementation a bit.

Specifically I think that, rather than add complexity to the `Badge` component by providing an optional `tooltip` property, we should instead update `Badge` to make it composable so we can wrap it in `Tooltip`, and then render the `Tooltip` around the `Badge` in the new `NewMetricBadge` component. That keeps `Badge` simpler and more flexible.

In other words, as per the [documentation](https://v4.mui.com/components/tooltips/#custom-child-element), `Badge` would need to be updated to use `forwardRef()` and to spread its props into its `span` component.

Then, `NewMetricBadge` could be implemented something like this:

```js
function NewMetricBadge( { tooltipTitle, learnMoreLink } ) {
	return (
		<Tooltip
			title={
				<Fragment>
					<p>{ tooltipTitle }</p>
					<Link href={ learnMoreLink } external>
						{ __( 'Learn more', 'google-site-kit' ) }
					</Link>
				</Fragment>
			}
			interactive
		>
			<Badge
				className="some-new-classname"
				label={ __( 'New', 'google-site-kit' ) }
			/>
		</Tooltip>
	);
}
```

How does that sound?

**Comment by techanvil** [comment]
Hi @sashadoes, thanks for the update, however there are still some issues to address in the IB. I'll go through them in turn.

> Inside 

There's a rogue word "Inside" that has been left in by accident.

>    * Introduce a new `tooltipTitle ` prop to configure tooltip content

This is fine, but you should also include a mention of the proposed `learnMoreLink` prop for the link URL.

>    * Use `Tooltip.js` to render a configurable tooltip inside `Badge` with the content

The `Tooltip` will now be rendered _around_ the `Badge` component, not inside it. Also, on a stylistic note, it's more conventional to describe a component as "the `Tooltip`" or "the `Tooltip` component" rather than `Tooltip.js` as you have here.

>    * Use `Link.js` to render a link in the tooltip as per AC

Again, this would be more conventionally described as "Use the `Link` component".

>    * Use `forwardRef()` as per the [documentation](https://v4.mui.com/components/tooltips/#custom-child-element), in Badge to spread its props into its span component

Super trivial and I'd usually just fix it myself but `Badge` and `span` here are missing the backticks to format them :)

> NewMetricBadge Eg: 
> ```js
> function NewMetricBadge( { tooltipTitle, learnMoreLink } ) {
> 	return (
> 		<Tooltip
> 			title={
> 				<Fragment>
> 					<p>{ tooltipTitle }</p>
> 					<Link href={ learnMoreLink } external>
> 						{ __( 'Learn more', 'google-site-kit' ) }
> 					</Link>
> 				</Fragment>
> 			}
> 			interactive
> 		>
> 			<Badge
> 				className="some-new-classname"
> 				label={ __( 'New', 'google-site-kit' ) }
> 			/>
> 		</Tooltip>
> 	);
> }
> ```

As a rule we don't include full code snippets like this in an IB. However, seeing as I did paste that in the comment below, it would be appropriate to link to the comment as an example, e.g. "For a bit more context, and a concrete example on how the `NewMetricBadge` component could look, see this [comment](https://github.com/google/site-kit-wp/issues/6552#issuecomment-1474183079) below".

> * Introduce a new "some-new-classname" and apply styles to `Badge` as per [Figma designs](https://www.figma.com/file/0UCIE3hPeiu4NYdpAODASd/GA4-Updates?node-id=171-3164&t=umqvQCOYgQI7xYXD-4) 

Without the inline example, 'Introduce a new "some-new-classname"' won't make so much sense here, so I'd suggest changing this to "Introduce a new class".

> * For `Sessions` and `Engaged` metrics add the `Badge` component in `/assets/js/components/ReportTable.js` after `{ title }` for the appropriate titles

The **Engaged Sessions** metric has actually been updated to **Engagement Rate**. I've updated the AC, but this point in the IB also needs to be updated. It would also be good to add a little more detail about the `ReportTable` component - this will also need a new `badge` prop, and it would be good to update the Storybook story for `ReportTable` to include an example of the `badge` prop being used.

It would also be useful to add some more Storybook requirements. We could add a story for the `NewMetricsBadge` component, and we should also add variants for the **Search Funnel** and **Popular Pages** widget stories which show the **New** badges.

With these additions, I'd suggest the estimate should probably be increased too.

**Comment by techanvil** [comment]
Hi @sashadoes, thanks for updating the IB. It still needed some changes, and I've gone ahead and updated it myself.

The most important changes made are that the widgets to be modified should be the GA4 versions, and we shouldn't hardcode exceptions for the metrics into `ReportTable`, rather we should keep this component generic and pass the badge prop in from its parent component. There are also some grammatical and other tweaks.

Please review the changes I've made and if you're happy with them we can move this to the EB.

**Comment by sashadoes** [comment]
Thanks @techanvil , moving it to IBR now 🙂

**Comment by sashadoes** [comment]
Thanks @techanvil , moving it to EB now.

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@kuasha420 I have a few observations:

1. For the three tooltips, the `New` text does not align up with the title.  It's only slightly off but you can see the difference when you look at the figma designs.  Note: I think this is because the titles on my test site are 14px whereas it is 12px on figma. So, this is probably why they do not look aligned.  Also, the area of the new badge is `41.25x24` on my test site and `42x24` on figma designs so that could also be the cause.  

2. The `Engagement Rate` badge is not aligned to the title and appears on the line below. On tablets, the issue is worse because the Sessions appears also on the line below and  `Engagement Rate` expands to three lines. 

3. The `new` badge only appears on the `Conversions` but when I tap on mobile, the tooltip does not appear. From memory, I don't think we usually have tooltips on mobile, so wanted to check this.  For note: the `Sessions` and `Engagement Rate` data does not appear in the widget on mobile so the new badges do not appear. 

The AC says:

> When hovering over (tapping on mobile) a "New" badge, a tooltip should appear like in the Figma design.

See the screenshots below:

<details>
 <summary>Screenshots</summary>

**Issue 1:**
![image](https://user-images.githubusercontent.com/73545194/229098528-30a4e742-e2bb-44d6-9f73-59215abaa7a0.png)

**Issue 2:**
![image](https://user-images.githubusercontent.com/73545194/229101835-012346ce-96b6-46a0-99aa-1b148615bfce.png)
![image](https://user-images.githubusercontent.com/73545194/229101715-ad9388fb-4837-4354-bb5e-0b85b7b0f678.png)

</details>

**Verified**

_Conversions:_ 
- The new badge appears on the conversions title. 
- The copy is as per Figma designs
- The `learn more` link points to https://support.google.com/analytics/answer/9267568

_Sessions:_
- The new badge appears on the sessions title. 
- The copy is as per Figma designs
- The `learn more` link points to https://support.google.com/analytics/answer/9191807

_Engagement Rate:_
- The new badge appears on the engagement rate title. 
- The copy is as per Figma designs
- The `learn more` link points to https://support.google.com/analytics/answer/12195621

**Comment by aaemnnosttv** [comment]
> 2\. The `Engagement Rate` badge is not aligned to the title and appears on the line below. On tablets, the issue is worse because the Sessions appears also on the line below and  `Engagement Rate` expands to three lines.

The appearance in smaller viewports could be better. We should sync with UX about this, it could be that moving the badge before the column name is all we need to do – it wouldn't fix the wrapping but it would keep the badge in a consistent place and the position of the title relative to the data below should be more consistent as well.

> 3\. From memory, I don't think we usually have tooltips on mobile, so wanted to check this

We don't have feature tours on mobile but IIRC tooltips should still work.

**Comment by kuasha420** [comment]
@wpdarren thank you both for the observations. I'll create a follow up PR that fixes 1 and 3, and we can keep 2 as a known issue until a decision is reached and fix it in a follow up issue.

**Comment by aaemnnosttv** [comment]
Back to you @wpdarren for another pass 👍

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified**

The issue with the tooltip on mobile (point 3 above) is fixed but still feel issue 1 is a problem.  I am going to move this forward and include a note for issue 1 and 2 on my comment above as known issues on the bug bush board because they will come up during the teams testing. 

_Conversions:_ 
- The new badge appears on the conversions title. 
- The copy is as per Figma designs
- The `learn more` link points to https://support.google.com/analytics/answer/9267568

_Sessions:_
- The new badge appears on the sessions title. 
- The copy is as per Figma designs
- The `learn more` link points to https://support.google.com/analytics/answer/9191807

_Engagement Rate:_
- The new badge appears on the engagement rate title. 
- The copy is as per Figma designs
- The `learn more` link points to https://support.google.com/analytics/answer/12195621

**Comment by aaemnnosttv** [comment]
Thanks @wpdarren – I've flagged the alignment issue with Sigal.

We can address this and any other details in follow-up issues 👍
